src/crafting.cpp

```diff
@@ -24,6 +24,28 @@
 #include <math.h>    //sqrt
 #include <algorithm> //std::min
 
+enum usage {
+	use_from_map = 1,
+	use_from_player = 2,
+	use_from_both = 1 | 2,
+	use_from_none = 4,
+	cancel = 5
+};
+
+struct comp_selection {
```

Comment:
This looks as if it's the same as the `tool_selection` only with different `comp` type. It could be made into a templated class:

```C++
template<typename CompType>
struct comp_selection {
    usage use_from;
    CompType comp;
};
using item_selection = comp_selection<item_comp>;
using tool_selection = comp_selection<tool_comp>;
```